File: src/friday/progress.py
from __future__ import annotations

from copy import deepcopy
from datetime import datetime
from pathlib import Path
from typing import Any

from agent_core import RunContext

PROGRESS_ARTIFACT = "friday.progress"
PROGRESS_HEADING = "## Current Session Progress"
STEP_STATUSES = {"pending", "in_progress", "completed", "blocked"}


def current_progress(context: RunContext) -> dict[str, Any]:
    value = context.artifacts.get(PROGRESS_ARTIFACT, {})
    return deepcopy(value) if isinstance(value, dict) else {}


def begin_progress(context: RunContext, request: str, *, mode: str, continuation: bool = False) -> dict[str, Any]:
    previous = current_progress(context)
    if not continuation or not previous:
        state = {
            "objective": request.strip(),
            "latest_request": request.strip(),
            "mode": mode,
            "status": "working",
            "steps": [],
            "next_action": "",
            "verification": {},
        }
    else:
        state = previous
        state["status"] = "working"
        state["next_action"] = ""
    return _store(context, state)


def update_plan(
    context: RunContext,
    plan: list[dict[str, Any]],
    *,
    objective: str = "",
    explanation: str = "",
    next_action: str = "",
) -> dict[str, Any]:
    steps = _validate_plan(plan)
    state = current_progress(context)
    if not state:
        raise RuntimeError("No active Friday progress state.")
    if objective.strip():
        state["objective"] = objective.strip()
    state["steps"] = steps
    state["status"] = "working"
    state["next_action"] = next_action.strip()
    return _store(context, state, explanation=explanation.strip())


def finish_progress(context: RunContext, loop_status: str, verifications: list[dict[str, Any]]) -> dict[str, Any]:
    state = current_progress(context)
    if not state:
        return {}
    last = verifications[-1] if verifications else {}
    state["verification"] = {
        key: last[key]
        for key in ("attempt", "stop_reason", "verdict")
        if key in last
    }
    if loop_status == "done":
        state["status"] = "done"
        state["next_action"] = ""
        state["steps"] = [{**step, "status": "completed"} for step in state.get("steps", [])]
    elif loop_status == "needs_approval":
        state["status"] = "waiting"
        state["next_action"] = "Choose whether to approve, allow for this session, reject, or provide guidance."
    else:
        state["status"] = "blocked"
        state["next_action"] = str(last.get("next_check") or last.get("feedback") or loop_status).strip()
    return _store(context, state)


def restore_progress(context: RunContext, value: Any) -> dict[str, Any]:
    if not isinstance(value, dict) or not str(value.get("objective") or "").strip():
        return {}
    try:
        steps = _validate_plan(value.get("steps") if isinstance(value.get("steps"), list) else [])
    except ValueError:
        steps = []
    status = str(value.get("status") or "working")
    if status not in {"blocked", "done", "waiting", "working"}:
        status = "working"
    state = {
        "objective": str(value.get("objective") or "").strip(),
        "latest_request": str(value.get("latest_request") or "").strip(),
        "mode": "goal" if value.get("mode") == "goal" else "normal",
        "status": status,
        "steps": steps,
        "next_action": str(value.get("next_action") or "").strip(),
        "verification": dict(value.get("verification") or {}) if isinstance(value.get("verification"), dict) else {},
        "updated": str(value.get("updated") or ""),
    }
    context.artifacts[PROGRESS_ARTIFACT] = state
    return current_progress(context)


def append_progress_checkpoint(context: RunContext) -> None:
    state = current_progress(context)
    if state:
        collections = [context.messages]
        scoped = context.get_messages()
        if scoped is not context.messages:
            collections.append(scoped)
        for messages in collections:
            if messages and is_progress_checkpoint(messages[-1]):
                messages.pop()
        context.add_message("assistant", progress_checkpoint(state), friday_progress=True)


def is_progress_checkpoint(message: dict[str, Any]) -> bool:
    if message.get("friday_progress"):
        return True
    # Sessions saved before the metadata flag existed carry only the heading.
    return str(message.get("content") or "").startswith(PROGRESS_HEADING)


def progress_checkpoint(state: dict[str, Any]) -> str:
    lines = [
        PROGRESS_HEADING,
        f"Objective: {_clip(str(state.get('objective') or ''), 1000)}",
        f"Mode: {state.get('mode', 'normal')}",
        f"Status: {state.get('status', 'working')}",
    ]
    latest = str(state.get("latest_request") or "").strip()
    if latest and latest != state.get("objective"):
        lines.append(f"Latest user update: {_clip(latest, 1000)}")
    steps = state.get("steps", [])
    if steps:
        lines.append("Plan:")
        marks = {"pending": "[ ]", "in_progress": "[>]", "completed": "[x]", "blocked": "[!]"}
        lines.extend(f"- {marks[step['status']]} {_clip(step['step'], 500)}" for step in steps)
    if state.get("next_action"):
        lines.append(f"Next action: {_clip(str(state['next_action']), 1000)}")
    return "\n".join(lines)


def progress_line(state: dict[str, Any]) -> str:
    if not state:
        return "No active session progress."
    steps = state.get("steps", [])
    completed = sum(step.get("status") == "completed" for step in steps)
    count = f" | {completed}/{len(steps)} steps" if steps else ""
    next_action = f" | next: {state['next_action']}" if state.get("next_action") else ""
    return f"[{state.get('status', 'working')}] {state.get('objective', '')}{count}{next_action}"


def _validate_plan(plan: list[dict[str, Any]]) -> list[dict[str, str]]:
    if len(plan) > 12:
        raise ValueError("Plan supports at most 12 steps.")
    steps: list[dict[str, str]] = []
    for item in plan:
        if not isinstance(item, dict):
            raise ValueError("Each plan item must be an object.")
        step = str(item.get("step") or "").strip()
        status = str(item.get("status") or "").strip()
        if not step or status not in STEP_STATUSES:
            raise ValueError("Plan items require step and status=pending|in_progress|completed|blocked.")
        steps.append({"step": step, "status": status})
    if sum(step["status"] == "in_progress" for step in steps) > 1:
        raise ValueError("At most one plan step can be in_progress.")
    return steps


def _store(context: RunContext, state: dict[str, Any], *, explanation: str = "") -> dict[str, Any]:
    state = deepcopy(state)
    state["updated"] = datetime.now().isoformat(timespec="seconds")
    context.set_artifact(PROGRESS_ARTIFACT, state)
    context.emit("progress.updated", category="progress", data={**state, "explanation": explanation})
    workspace = context.metadata.get("workspace")
    session_id = context.metadata.get("session_id")
    if workspace and session_id:
        from friday.state import save_progress

        save_progress(Path(str(workspace)), str(session_id), state)
    return current_progress(context)


def _clip(text: str, limit: int) -> str:
    return text if len(text) <= limit else text[: limit - 3] + "..."

File: packages/harness/src/progress.ts
import type { RunContext } from 'friday-agent-core'

import { localTimestamp } from './time.js'

export type StepStatus = 'pending' | 'in_progress' | 'completed' | 'blocked'
export type ProgressStep = { step: string; status: StepStatus }
export type ProgressState = {
  objective: string
  latest_request: string
  mode: 'normal' | 'goal'
  status: 'working' | 'waiting' | 'blocked' | 'done'
  steps: ProgressStep[]
  next_action: string
  verification: Record<string, unknown>
  updated: string
}

const ARTIFACT = 'friday.progress'
const STEP_STATUSES = new Set<StepStatus>(['pending', 'in_progress', 'completed', 'blocked'])

export function currentProgress(context: RunContext): ProgressState | undefined {
  const value = context.artifacts[ARTIFACT]
  return isProgress(value) ? structuredClone(value) : undefined
}

export function beginProgress(
  context: RunContext,
  request: string,
  mode: 'normal' | 'goal' = 'normal',
  continuation = false
): ProgressState {
  const previous = currentProgress(context)
  const objective = request.trim()
  const state: ProgressState = continuation && previous ? {
    ...previous,
    latest_request: objective || previous.latest_request,
    status: 'working',
    next_action: ''
  } : {
    objective,
    latest_request: objective,
    mode,
    status: 'working',
    steps: [],
    next_action: '',
    verification: {},
    updated: ''
  }
  return store(context, state)
}

export function updatePlan(
  context: RunContext,
  value: { plan: unknown; objective?: unknown; explanation?: unknown; next_action?: unknown }
): ProgressState {
  const state = currentProgress(context)
  if (!state) throw new Error('No active Friday progress state.')
  const steps = validatePlan(value.plan)
  const objective = optionalText(value.objective, 'objective', 2_000)
  const explanation = optionalText(value.explanation, 'explanation', 2_000)
  const nextAction = optionalText(value.next_action, 'next_action', 2_000)
  return store(context, {
    ...state,
    ...(objective ? { objective } : {}),
    steps,
    status: 'working',
    next_action: nextAction
  }, explanation)
}

export function resumeProgress(context: RunContext): ProgressState | undefined {
  const state = currentProgress(context)
  return state ? store(context, { ...state, status: 'working', next_action: '' }) : undefined
}

export function recordVerificationProgress(context: RunContext, value: Record<string, unknown>): ProgressState | undefined {
  const state = currentProgress(context)
  if (!state) return undefined
  const verification = Object.fromEntries(
    ['attempt', 'stop_reason', 'verdict'].flatMap(key => value[key] === undefined ? [] : [[key, value[key]]])
  )
  return store(context, { ...state, verification })
}

export function finishProgress(
  context: RunContext,
  status: 'done' | 'waiting' | 'blocked',
  verification: Record<string, unknown> = {}
): ProgressState | undefined {
  const state = currentProgress(context)
  if (!state) return undefined
  const summary = Object.fromEntries(
    ['attempt', 'stop_reason', 'verdict'].flatMap(key => verification[key] === undefined ? [] : [[key, verification[key]]])
  )
  if (status === 'done') {
    return store(context, {
      ...state,
      status,
      next_action: '',
      steps: state.steps.map(step => ({ ...step, status: 'completed' })),
      verification: summary
    })
  }
  return store(context, {
    ...state,
    status,
    next_action: status === 'waiting'
      ? 'Choose whether to approve, allow for this session, reject, or provide guidance.'
      : state.next_action || 'Provide guidance or revise the request.',
    verification: summary
  })
}

export function restoreProgress(context: RunContext, value: unknown, emit = false): ProgressState | undefined {
  if (!isObject(value) || typeof value.objective !== 'string' || !value.objective.trim()) {
    delete context.artifacts[ARTIFACT]
    if (emit) context.emit('progress.updated', 'progress', {})
    return undefined
  }
  let steps: ProgressStep[] = []
  try { steps = validatePlan(value.steps) } catch {}
  const state: ProgressState = {
    objective: value.objective.trim(),
    latest_request: typeof value.latest_request === 'string' ? value.latest_request.trim() : '',
    mode: value.mode === 'goal' ? 'goal' : 'normal',
    status: ['working', 'waiting', 'blocked', 'done'].includes(String(value.status))
      ? value.status as ProgressState['status']
      : 'working',
    steps,
    next_action: typeof value.next_action === 'string' ? value.next_action.trim() : '',
    verification: isObject(value.verification) ? structuredClone(value.verification) : {},
    updated: typeof value.updated === 'string' ? value.updated : ''
  }
  context.artifacts[ARTIFACT] = structuredClone(state)
  if (emit) context.emit('progress.updated', 'progress', state)
  return structuredClone(state)
}

function validatePlan(value: unknown): ProgressStep[] {
  if (!Array.isArray(value)) throw new Error('plan must be an array.')
  if (value.length > 12) throw new Error('Plan supports at most 12 steps.')
  const steps = value.map((item, index) => {
    if (!isObject(item)) throw new Error(`plan[${index}] must be an object.`)
    const step = typeof item.step === 'string' ? item.step.trim() : ''
    const status = item.status
    if (!step || step.length > 500 || !STEP_STATUSES.has(status as StepStatus)) {
      throw new Error('Plan items require a short step and status=pending|in_progress|completed|blocked.')
    }
    return { step, status: status as StepStatus }
  })
  if (steps.filter(step => step.status === 'in_progress').length > 1) {
    throw new Error('At most one plan step can be in_progress.')
  }
  return steps
}

function store(context: RunContext, value: ProgressState, explanation = ''): ProgressState {
  const state = { ...structuredClone(value), updated: localTimestamp() }
  context.artifacts[ARTIFACT] = state
  context.emit('progress.updated', 'progress', { ...state, explanation })
  return structuredClone(state)
}

function optionalText(value: unknown, name: string, maximum: number): string {
  if (value === undefined || value === null) return ''
  if (typeof value !== 'string') throw new Error(`${name} must be text.`)
  const text = value.trim()
  if (text.length > maximum) throw new Error(`${name} must be at most ${maximum} characters.`)
  return text
}

function isProgress(value: unknown): value is ProgressState {
  return isObject(value) && typeof value.objective === 'string' && Array.isArray(value.steps)
}

function isObject(value: unknown): value is Record<string, unknown> {
  return !!value && typeof value === 'object' && !Array.isArray(value)
}
